extensions/amp-ad/0.1/amp-ad-3p-impl.js

```diff
@@ -176,24 +179,24 @@ export class AmpAd3PImpl extends AMP.BaseElement {
 
   /** @override */
   layoutCallback() {
-    if (!this.iframe_) {
-      user().assert(!this.isInFixedContainer_,
-          '<amp-ad> is not allowed to be placed in elements with ' +
-          'position:fixed: %s', this.element);
-      incrementLoadingAds(this.win);
-      return getAdCid(this).then(cid => {
-        const opt_context = {
-          clientId: cid || null,
-          container: this.container_ ? this.container_.tagName : null,
-        };
-        this.iframe_ = getIframe(this.element.ownerDocument.defaultView,
-            this.element, null, opt_context);
-        this.apiHandler_ = new AmpAdApiHandler(
-            this, this.element, this.boundNoContentHandler_);
-        return this.apiHandler_.startUp(this.iframe_, true);
-      });
+    if (this.layoutPromise_) {
+      return this.layoutPromise_;
     }
-    return this.loadPromise(this.iframe_);
+    user().assert(!this.isInFixedContainer_,
+        '<amp-ad> is not allowed to be placed in elements with ' +
+        'position:fixed: %s', this.element);
+    incrementLoadingAds(this.win);
+    return this.layoutPromise_ = getAdCid(this).then(cid => {
+      const opt_context = {
+        clientId: cid || null,
+        container: this.container_ ? this.container_.tagName : null,
```

Comment:
nit: since we're only using the container's tagname, we could probably just store the tagname in the first place (`getAdContainer`). This would avoid this null check here.